File: Data Structures & Algorithms/two-integer-sum/submission-5.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        HashMap<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            if (map.containsKey(nums[i])) {
                // we have a duplicate
                if (target == nums[i] * 2) {
                    int[] arr = new int[2];
                    arr[0] = Math.min(map.get(nums[i]), i);
                    arr[1] = Math.max(map.get(nums[i]), i);
                    return arr;
                }
            }
            map.put(nums[i], i);
        }

        for (int num : nums) {
            if (map.containsKey(target - num) && map.get(target - num) != map.get(num)) {
                int[] arr = new int[2];
                arr[0] = Math.min(map.get(target - num), map.get(num));
                arr[1] = Math.max(map.get(target - num), map.get(num));
                return arr;
            }
        }

        return new int[2];
    }
}

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-5.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if len(s) == 0:
            return 0

        r = 0
        l = 0
        best = 1
        mp = {}
        while r < len(s):
            if s[r] in mp:
                l = max(mp[s[r]] + 1, l)

            mp[s[r]] = r
            best = max(best, r - l + 1)
            r += 1
                

        return best